src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -38,29 +38,11 @@ impl JoinOrderTree {
         }
     }
 
-    pub(super) fn iter(&self) -> JoinOrderTreeIterator {
-        JoinOrderTreeIterator { stack: vec![self] }
-    }
-}
-
-pub(super) struct JoinOrderTreeIterator<'a> {
-    stack: Vec<&'a JoinOrderTree>,
-}
-
-impl<'a> Iterator for JoinOrderTreeIterator<'a> {
-    type Item = usize;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        while let Some(node) = self.stack.pop() {
-            match node {
-                JoinOrderTree::Relation(id) => return Some(*id),
-                JoinOrderTree::Join(left, right) => {
-                    self.stack.push(left);
-                    self.stack.push(right);
-                }
-            }
+    fn iter(&self) -> Box<dyn Iterator<Item = usize> + '_> {
+        match self {
+            JoinOrderTree::Relation(id) => Box::new(std::iter::once(*id)),
+            JoinOrderTree::Join(left, right) => Box::new(left.iter().chain(right.iter())),
```

Comment:
Are we boxing because the two branches return slightly different opaque types?

Reply:
Yeah, ideally I wanted to return `impl Iterator<Item = usize>`, but since `iter::once` is a different type from `iter::chain`, I needed to return `dyn Iterator<Item = usize>` instead. And since `dyn Iterator` is not sized it needs the `Box` around it